### PR TITLE
Compatibility with github v3 api

### PR DESCRIPTION
**Pull request by MichaelMayorov**
Support for GitHub v3 API.

Changes:
1. gh_values.py -- contains urls and templates for all requests to API, to change default repo and user, just edit "gh_repo"  and "gh_user"
2. Degraded performance for "sympy-bot list" and "sympy-bot list-numbers" commands, because "mergeable", "email" and "name" were removed from server response. So, sympy-bot will do 2*n + 1 requests to GitHub API
3. Cosmetic changes in some code parts.

Since code not covered by any tests, I ask community to test bot functional(I'll try to add some tests later)

### CODE REVIEW

File: sympy-bot [hunk 1]
```diff
@@ -93,7 +93,7 @@ Commands:
     proto = options.proto if options.proto != default_proto else default_proto
 
     options.user = config.get("user")
-    options.token = config.get("token")
```
Comment (Krastanov) [inline sympy-bot:96]: Tokens do not exist anymore in the api, is this right?

File: sympy-bot [hunk 2]
```diff
@@ -267,19 +267,20 @@ def review(n, config, **kwargs):
         config.python3 = True
     # If set Python 2 interpreter with -i and set -3, raise error
     elif config.python3:
-        raise ValueError('Python 2 interpreter passed with -3 option')
+        raise ValueError("Python 2 interpreter passed with -3 option")
```
Comment (Krastanov) [inline sympy-bot:273]: It is good idea not to change the code like that. This was correct code that complies with PEP8 and these changes only pollute git history.

File: sympy-bot [hunk 3]
```diff
@@ -267,19 +267,20 @@ def review(n, config, **kwargs):
...
+    repo = pull["head"]['repo']["html_url"]
     repo = repo.replace(default_proto, proto) if proto != default_proto else repo
-    branch = pull['head']['ref']
-    user = pull["user"]
-    author = '"%s" <%s>' % (pull["user"].get("name", "unknown"),
-                            pull["user"].get("email", ""))
+    branch = pull["head"]["ref"]
+    user = pull["head"]["user"]
+    user_info = github_get_user_info(username)
+    author = "\"%s\" <%s>" % (user_info.get("name", "unknown"),
```
Comment (Krastanov) [inline sympy-bot:285]: Again, it is not necessary to change the quote type. Here even more so, as the previous string with the nested quotes was more readable.

File: gh_values.py [hunk 4]
```diff
@@ -0,0 +1,11 @@
+# This file contains predefined values for communicating with GitHub
+# If you want to run bot on another repo, then replace "gh_user" and "gh_repo" with
+# yours values
+gh_api_url = "https://api.github.com"
+gh_user = "sympy"
+gh_repo = "sympy"
```
Comment (Krastanov) [inline gh_values.py]: I think that this should be loaded by in the configuration file parser. Is the configuration file still usable?

File: gh_values.py [hunk 5]
```diff
@@ -0,0 +1,11 @@
+# This file contains predefined values for communicating with GitHub
+# If you want to run bot on another repo, then replace "gh_user" and "gh_repo" with
+# yours values
+gh_api_url = "https://api.github.com"
+gh_user = "sympy"
```
Comment (Krastanov) [inline gh_values.py]: what is the difference between this and `option.user`?

Comment (Krastanov) [inline gh_values.py]: It has nothing to do with `option.user`. I think that this needs another name. How was the bot forced to use other repos before these changes?

File: utils.py [hunk 6]
```diff
@@ -281,9 +284,8 @@ def get_password():
     if config.user:
         username = config.user
 
-        if config.token:
-            username = username + '/token'
-            password = config.token
+        if config.password:
+            password = config.password
```
Comment (Krastanov) [inline utils.py:294]: This means that you are saving an unencrypted password. I think that this must be removed.

**Comment by Krastanov** [comment]
I do not think that it is save to change the argument lists of all these functions. It does not seem necessary (it is not imposed by the api changes) and it may cause problems in other untested parts of the codebase. Moreover, this change is linked to the fact that the repo is not accessible from the config file anymore.

**Comment by Krastanov** [comment]
The question is how quickly we want to fix the bot. If it is absolutely urgent this seems good enough. On the other hand it seems that it breaks parts of the config file and it changes the semantics of many functions (basically it seems more invasive than necessary).

Anyway, I will probably start using it as it is. Thanks for all the work!

File: utils.py [hunk 7]
```diff
@@ -1,18 +1,10 @@
...
-import subprocess
-import os
-from os.path import expandvars
-import urllib2
-import base64
-from urllib import urlencode
-import time
-import sys
-import re
-from getpass import getpass
```
Comment (scolobb) [inline utils.py:12]: I'm not sure that using fully qualified names is so much better than doing `from ... import ...`.  I'd recommend following the strategy of only doing the modifications which are really needed only, but others may disagree.

File: sympy-bot (hunk 1 above)
Comment (MichaelMayorov) [inline sympy-bot:96]: Yes, regards this http://developer.github.com/v3/#authentication
GitHub supports only plaintext login\password and OAuth2 for auth

File: utils.py (hunk 6 above)
Comment (scolobb) [inline utils.py:294]: When I used to run `sympy-bot` some time ago, it used to tell me that my password was to be stored in a Python variable for the period the bot runs.  If I am not mistaken then, perhaps, what happens here is not so much different from the previous state of affairs.

File: sympy-bot (hunk 2 above)
Comment (MichaelMayorov) [inline sympy-bot:273]: I agreed that in this pull I shouldn't do such kind of thing, but look in another places of code. It mixed up single quotes(') and double quotes(") everywhere. So, I thought that using only double quotes will make code more consistent.

File: utils.py (hunk 6 above)
Comment (MichaelMayorov) [inline utils.py:294]: @Krastanov If you don't want to store your password, then don't do it. You could type password every time when bot started.

File: sympy-bot (hunk 2 above)
Comment (scolobb) [inline sympy-bot:273]: There indeed was a mix of different quoting styles without much motivation, like in lines 272 and 273, where first single quotes and the double quotes are used, and it's never bad to fix such problems.  However, you absolutely right in that such changes do not belong to this pull request.

File: gh_values.py (hunk 5 above)
Comment (MichaelMayorov) [inline gh_values.py]: I think option.user should be removed, because using config file for getting this value is much comfortable as for user as for developer. The main reason - dynamic links and templates for URLs. Just looks how URLs created in old version

File: utils.py (hunk 7 above)
Comment (Krastanov) [inline utils.py:12]: I was thinking about making the same remark...

File: utils.py (hunk 6 above)
Comment (Krastanov) [inline utils.py:294]: There is a big difference: storing something in volatile memory is not as dangerous as storing it to disk. This is why every desktop shell implements some form of keyring and this is why Oauth and tokens exist.

I do not consider this a blocker for a merge and you can leave it as it is, however it is bad design. I do agree that we need not preoccupy ourselves with security issues.

File: gh_values.py (hunk 5 above)
Comment (Krastanov) [inline gh_values.py]: What do you call a config file? Changing this `gh_values` will work, however it is not a config file. The bot already supports a config file in the `.sympy` directory.

File: gh_values.py (hunk 4 above)
Comment (MichaelMayorov) [inline gh_values.py]: If I understand you question right, you should see difference between user's(sympy-bot.conf) and bot's(gh_values.py) defined values. First stores user defined values such as login or password from your account. Second stores repo name, owner's login and other related values. So these files have different purposes.

File: utils.py (hunk 6 above)
Comment (scolobb) [inline utils.py:294]: Oh, storing an unencrypted password in persistent storage is definitely bad.  Indeed, security issues are lower priority now, but I'd still recommend just removing the storing feature altogether.

How critical is having persisted passwords?  I, for one, always type in my credentials each time I run `sympy-bot`.

File: gh_values.py (hunk 4 above)
Comment (Krastanov) [inline gh_values.py]: All of these options seem to have been available from the config file. For instance this is from the bot's help page:

```
  -R REPOSITORY, --repository=REPOSITORY
                        Change the default 'sympy/sympy' repository to
                        something else. This allows to use sympy-bot with
                        other projects.

```

File: utils.py (hunk 6 above)
Comment (MichaelMayorov) [inline utils.py:294]: I fully agreed with you, but GitHub doesn't leave me a choice(I'll try to add OAuth auth later) but for now it's only one way to use sympy-bot without annoying asking of password :-(

File: sympy-bot (hunk 2 above)
Comment (Krastanov) [inline sympy-bot:273]: Actually not having a persistent way to authenticate is a very big pain. Now that tokens are not available either you have to store your password (easy but not secure) or set up OAuth (complicated).

We can leave this as it is for the moment, however there should be definitely an issue filed against it.

File: utils.py (hunk 6 above)
Comment (Krastanov) [inline utils.py:294]: > I fully agreed with you, but GitHub doesn't leave me a choice(I'll try to add OAuth auth later) but for now it's only one way to use sympy-bot without annoying asking of password :-(

It is a pain right now. Do not dedicate too much time to this. What
you have done is already of great value; we can think about security
later (after all this is just a build-bot).

File: gh_values.py (hunk 4 above)
Comment (MichaelMayorov) [inline gh_values.py]: Yes, you could change default user/repo via this option, but if you don't want to do it each time, you could change "gh_user" and "gh_repo" in gh_values.py

**Comment by Krastanov** [comment]
@MichaelMayorov, by the way, do not feel pressured to respond right away to all the raised issues. Now that thanks to you we have again a working bot we can go more slowly.

File: gh_values.py (hunk 4 above)
Comment (Krastanov) [inline gh_values.py]: But it seems that previously this was also available in the config file. Moreover, at the moment it is actually impossible to use this option, because your code overrides it.

Another difference is that previously  `repository` contained `gh_user+'/'+gh_repo` but now you are dividing them up in gh_user and gh_repo.

Comment (MichaelMayorov) [inline gh_values.py]: > Moreover, at the moment it is actually impossible to use this option, because your code overrides it.

Thanks for testing! I reproduced this problem http://pastebin.com/xBBhjaCX so you're right. I'll fix this problem tomorrow.

Comment (MichaelMayorov) [inline gh_values.py]: BTW, I also plan to add tests to cover this and other options of bot

File: utils.py (hunk 6 above)
Comment (asmeurer) [inline utils.py:294]: How about we leave this for now, but remove it when we get OAuth workin?. I should note that the old way of storing a token was just as insecure as this, as the token allowed essentially the same access as a password.

File: gh_values.py (hunk 4 above)
Comment (asmeurer) [inline gh_values.py]: So I guess ideally the config file should override these values if given, and command line options should override both.

File: gh_values.py [hunk 8]
```diff
@@ -0,0 +1,11 @@
...
+gh_repo = "sympy"
+gh_pull_list_url = gh_api_url + "/repos" + "/" + gh_user + "/" + gh_repo + "/pulls" 
+gh_pull_template = gh_pull_list_url + "/%d"
+gh_user_info_template = gh_api_url + "/users/%s"
+gh_issue_comment_template = gh_api_url + "/repos" + "/" + gh_user + "/" + gh_repo + \
```
Comment (asmeurer) [inline gh_values.py]: I you're going to modify this, why not use new style string formatting?  That will make it easier to read both here anywhere it is used. Note that we don't have to worry about Python 2.5 support in SymPy bot, so that's not an issue.

File: gh_values.py (hunk 4 above)
Comment (Krastanov) [inline gh_values.py]: > So I guess ideally the config file should override these values if given, and command line options should override both.
> This was the original behavior. All that is needed is a copy/paste
> from the original code.

File: utils.py [hunk 9]
```diff
@@ -302,3 +304,33 @@ def get_password():
         password = get_password()
 
     return username, password
+
+def query2github(url, username="", password="", data=""):
+    """
+    Query github API,
+    if username and password presented, then query executed from user account
```
Comment (certik) [inline utils.py]: I would write "if username and password are presented, then the query is executed from the user account", but I'll leave it to some native speaker to verify the grammar (@asmeurer?).

**Comment by certik** [comment]
Hi Michael, thanks for updating the sympy-bot! I am ok with the changes.

File: utils.py (hunk 7 above)
Comment (certik) [inline utils.py:12]: In general, the best is to not touch the original cosmetic things like this and simply implement the needed functionality. And then send a different pull request with just the cosmetic things, since those are unrelated. I think I wrote most of these import lines (at least "git blame" suggests that), and I am personally fine either way, so whatever works for you is ok with me too.

File: utils.py (hunk 9 above)
Comment (asmeurer) [inline utils.py]: @certik what you wrote is correct.

File: utils.py (hunk 7 above)
Comment (MichaelMayorov) [inline utils.py:12]: @certik ok, I'll get back these changes, but even so, I think this code should be reorganized.
1. 

> from urllib import urlencode

Should be removed with pastehtml functionality, because not used.
2. For me, mess "from" and "import" in one code block is ugly
3. "os", "time", "sys", "re", "subprocess" could be written in one "import" line, that will improve readability

Comment (asmeurer) [inline utils.py:12]: It's actually common Python convention to have each module import on a separate line.

Comment (MichaelMayorov) [inline utils.py:12]: ok, I returned it now, you could check this out in 2 latest commits.

**Comment by MichaelMayorov** [comment]
All right, I restored "-R" option behaviour, but that code required testing.

**Comment by MichaelMayorov** [comment]
I have some free time on this week, which I can dedicate on implementing new features in sympy-bot. I don't hurry, but if someone helps me with review, I'll start working on new ideas immediately.

**Comment by scolobb** [comment]
That's just great!  Go ahead with the implementation and submit pull requests, I guess there will always be someone to help you either here or on the mailing list.

File: url_templates.py [hunk 10]
```diff
@@ -0,0 +1,17 @@
+class URLs(object):
+    """
+    This class contains URLs and templates which used in requests to GitHub API
```
Comment (certik) [inline url_templates.py:3]: which are used

**Comment by certik** [comment]
I am +1 to merge this as it is now. @asmeurer and others, are you ok with it too?

**Comment by asmeurer** [comment]
I haven't had a chance to test it, but if it works please do.

**Comment by Krastanov** [comment]
I have just started running this bot on one of the pull requests. The config file works nicely. I will wait to see that it actually uploads a comment and I will merge this if it is not already merged by then.

**Comment by Krastanov** [comment]
It works, thank you very much @MichaelMayorov. I am merging this.
